Sming/Components/crypto/include/Crypto/HashContext.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <axtls-8266/crypto/crypto.h>
+#include <WString.h>
+#include <Data/HexString.h>
+
+namespace Crypto
+{
+/**
+ * @brief
+ */
+struct Blob {
+	const void* data;
+	size_t length;
+
+	Blob(const void* data, size_t length) : data(data), length(length)
+	{
+	}
+
+	Blob(const String& str) : data(str.c_str()), length(str.length())
+	{
+	}
+};
+
+/**
+ * @brief Class template for a Hash value, essentially just an array of bytes
+ */
+template <size_t size> struct HashValue {
+	uint8_t data[size];
+
+	static constexpr size_t length = size;
+
+	bool copy(const void* src)
```

Comment:
Should be constructor, i.e.
```
auto hash = HashValue(data);
```
instead of
```
HashValue hash;
hash.copy(data)
```

Reply:
Adding constructors breaks POD and aggregate initialisation which we need for things like this:

```
static constexpr Crypto::MD5::Hash myHash PROGMEM = { ... };
```

I had this issue with FlashString.

Reply:
Ok, didn't know that.

Reply:
https://sming.readthedocs.io/en/latest/_inc/Sming/Components/FlashString/object.html#aggregate-initialization

Reply:
Too me more time than I'd care to admit working up a solid solution to that!

Reply:
Because of this particular issue, I added additional checks to the DEFINE macros in FlashString:

```
#define FSTR_CHECK_STRUCT(name)                                                                                        \
	static_assert(std::is_pod<decltype(name)>::value, "FSTR structure not POD");                                       \
	static_assert(offsetof(decltype(name), data) == sizeof(uint32_t), "FSTR structure alignment error");

```

Both of these issues caused me headaches. I think Crypto is simple enough we can omit this, but something to bear in mind just in case.

Reply:
Thanks for the link, I did read the cppreference page in the meantime. That would have thrown me off quite a while, too.

Reply:
NB. `std::array` doesn't have a constructor for exactly this reason.